clippy_lints/src/manual_clamp.rs

```diff
@@ -0,0 +1,733 @@
+    ///
+    /// ### Why is this bad?
+    /// clamp is much shorter, easier to read, and doesn't use any control flow.
+    ///
+    /// ### Known issue(s)
+    /// If the clamped variable is NaN this suggestion will cause the code to propagate NaN
+    /// rather than returning either `max` or `min`.
+    ///
+    /// `clamp` functions will panic if `max < min`, `max.is_nan()`, or `min.is_nan()`.
+    /// Some may consider panicking in these situations to be desirable, but it also may
+    /// introduce panicking where there wasn't any before.
+    ///
+    /// ### Examples
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// if input > max {
+    ///     max
+    /// } else if input < min {
+    ///     min
+    /// } else {
+    ///     input
+    /// }
+    /// # ;
+    /// ```
+    ///
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// input.max(min).min(max)
+    /// # ;
+    /// ```
+    ///
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// match input {
+    ///     x if x > max => max,
+    ///     x if x < min => min,
+    ///     x => x,
+    /// }
+    /// # ;
+    /// ```
+    ///
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// let mut x = input;
+    /// if x < min { x = min; }
+    /// if x > max { x = max; }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// # let (input, min, max) = (0, -2, 1);
+    /// input.clamp(min, max)
+    /// # ;
+    /// ```
+    #[clippy::version = "1.66.0"]
+    pub MANUAL_CLAMP,
+    complexity,
+    "using a clamp pattern instead of the clamp function"
+}
+impl_lint_pass!(ManualClamp => [MANUAL_CLAMP]);
+
+pub struct ManualClamp {
+    msrv: Option<RustcVersion>,
+}
+
+impl ManualClamp {
+    pub fn new(msrv: Option<RustcVersion>) -> Self {
+        Self { msrv }
+    }
+}
+
+type ExprRef<'tcx> = &'tcx Expr<'tcx>;
+
+#[derive(Debug)]
+struct ClampSuggestion<'tcx> {
+    params: InputMinMax<'tcx>,
+    span: Span,
+    make_assignment: Option<ExprRef<'tcx>>,
+}
+
+#[derive(Debug)]
+struct InputMinMax<'tcx> {
+    input: ExprRef<'tcx>,
+    min: ExprRef<'tcx>,
+    max: ExprRef<'tcx>,
+    is_float: bool,
+}
+
+impl<'tcx> LateLintPass<'tcx> for ManualClamp {
+    fn check_expr(&mut self, ctx: &LateContext<'tcx>, expr: ExprRef<'tcx>) {
+        if !meets_msrv(self.msrv, msrvs::CLAMP) {
+            return;
+        }
+        if !expr.span.from_expansion() {
+            check_anti_patterns(ctx, expr);
+        }
+    }
+    extract_msrv_attr!(LateContext);
+}
+
+fn check_anti_patterns<'tcx>(cx: &LateContext<'tcx>, expr: ExprRef<'tcx>) {
+    let suggestion_iter = is_if_elseif_else_pattern(cx, expr)
+        .or_else(|| is_max_min_pattern(cx, expr))
+        .or_else(|| is_call_max_min_pattern(cx, expr))
+        .or_else(|| is_match_pattern(cx, expr))
+        .or_else(|| is_if_elseif_pattern(cx, expr))
+        .into_iter()
+        .chain(is_two_if_pattern(cx, expr));
```

Comment:
Done, including the test

Reply:
`is_two_if_pattern` still needs to use `span_lint_hir_and_then` rather than `span_lint_and_then`, that'll make the following work:

```rust
#[allow(clippy::manual_clamp)]
if input < 10 { input = 10 }
if input > 20 { input = 20 }
```

Reply:
Done!